### PR TITLE
year_int_from_date_str in parse_date gem

### PR DESCRIPTION
**Pull request by ndushay**
## Why was this change made?

This gem contains string date parsing code for DLSS.  This particular PR ports over code from the stanford-mods gem for `#year_int_from_date_str` and for `year_str_valid?` methods from https://github.com/sul-dlss/stanford-mods/blob/master/lib/stanford-mods/date_parsing.rb (though in this context, the methods are in a module, not a class.)

In the short term, this gem will be used in dlme-transform in place of stanford-mods, which was added in PR sul-dlss/dlme-transform/pull/231.

In the near term, the code here will be expanded to cope with year ranges, with negative numbered years (beyond -999) and with years of fewer than 4 digits.  This will support DLME.

At some point in the future, we may want to refactor stanford-mods to use the methods here.

## Was the documentation (README, API, wiki, ...) updated?

README was updated.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,7 +25,31 @@ Or install it yourself as:
...
-TODO: Write usage instructions here
+ParseDate is a module with date string parsing methods.
+
+Include the desired functionality in the class of your choice:
+
+```
+require 'parse_date/int_from_string'
+
+class DummyClass
+  include ParseDate::IntFromString
```
Comment (jcoyne) [inline README.md]: This is an interesting choice.  Why make this a mixin rather than a plain class or using static methods?  This seems like it leads to mixing concerns in one class. 

I think I'd prefer something like: 
```ruby
ParseDate.year_int_from_date('12/25/00')
#=> 2000
```

Or 
```ruby
parser = ParseDate.new
parser.year_int_from_date('12/25/00')
#=> 2000
```

File: spec/parse_date/int_from_string_spec.rb [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+    'January 22th [1800]' => '1800',
+    'June the 12, 1794' => '1794',
+    'Mai 1st 1789' => '1789',
+    'March 22 d. 1794' => '1794',
+    'N. 7 1796' => '1796',
+    'N[ovember] 21st 1786' => '1786',
+    'Oct. the 2.d 1793' => '1793'
+  }
+  # example string as key, expected parsed value as value
+  specific_day_2_digit_year = {
```
Comment (jcoyne) [inline spec/parse_date/int_from_string_spec.rb]: These seem ambiguous.  What about 12/15/37,  Emperor Nero's birthday ?

File: spec/parse_date/int_from_string_spec.rb [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+    '1/2/79' => '1979',
+    '2/12/15' => '2015',
+    '6/11/99' => '1999',
+    '10/1/90' => '1990',
+    '10/21/08' => '2008',
```
Comment (jcoyne) [inline spec/parse_date/int_from_string_spec.rb]: The Games of the IV Olympiad began 4/27/08 (1908).

File: spec/parse_date/int_from_string_spec.rb [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+    '-999' => '-001',
+    '-914' => '-086',
+    '-18' => '-982',
+    '-1' => '-999',
+    '0' => '0000',
+    '5' => '0005',
+    '33' => '0033',
+    '945' => '0945'
+  }
+  bc_dates_to_int = {
```
Comment (jcoyne) [inline spec/parse_date/int_from_string_spec.rb]: how about "BCE", "bce" (as used https://www.oxfordreference.com/view/10.1093/acref/9780191735370.timeline.0001)

File: spec/parse_date/int_from_string_spec.rb [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+  }
+  # example string as key, expected parsed value as value
+  multiple_years_4_digits_once = {
+    '1918-20' => ['1918', '1919', '1920'], # vs. 1961-04
+    '1965-8' => ['1965', '1966', '1967', '1968'], # revs
+    '[1846-51]' => ['1846', '1847', '1848', '1849', '1850', '1851']
+  }
+  # example string as key, expected parsed value as value
+  decade_only_4_digits = {
+    'early 1890s' => ['1890', '1891', '1892', '1893', '1894', '1895', '1896', '1897', '1898', '1899'],
```
Comment (jcoyne) [inline spec/parse_date/int_from_string_spec.rb]: 1899  is not early 1890s

File: README.md [hunk 6]
```diff
@@ -25,7 +25,31 @@ Or install it yourself as:
...
+end
+```
+
+And then you may use the public methods as you like:
```
Comment (jcoyne) [inline README.md]: Can you add  `year_int_valid?`

File: spec/parse_date/int_from_string_spec.rb [hunk 7]
```diff
@@ -0,0 +1,578 @@
...
+      it "#{expected} for #{example}" do
+        expect(DummyClass.new.year_int_valid?(example)).to eq expected
+      end
+    end
+    it 'true for 0000' do
+      expect(DummyClass.new.year_int_valid?(0000)).to eq true
+    end
+  end
+
+  context '#sortable_year_for_yyyy' do
```
Comment (jcoyne) [inline spec/parse_date/int_from_string_spec.rb]: why bother testing this private method if all it's public usages are already tested via the `year_int_from_date_str` tests?

File: README.md (hunk 1 above)
Comment (ndushay) [reply README.md]: it felt clunky;  that was how I did it in `Stanford::Mods::DateParsing`.   I can change it, sure.  I was thinking about what was optimal and maybe landed in an unusual place.

File: spec/parse_date/int_from_string_spec.rb (hunk 2 above)
Comment (ndushay) [reply spec/parse_date/int_from_string_spec.rb]: At the time, I was basing it on actual date strings in the SDR and what we needed for SearchWorks to be able to integration some additional collections with weird dates.   This is coming from stanford-mods;  I haven't changed the logic here -- I'm just porting functionality to a more useful place.

File: spec/parse_date/int_from_string_spec.rb (hunk 3 above)
Comment (ndushay) [reply spec/parse_date/int_from_string_spec.rb]: imperfect?  yes.  Same comment as above applies.   I think we're unlikely to see "4/27/08" as a string and have it refer to 1908.  I'm not sure when the slash syntax came into favor, but I suspect it is much more recent.

File: spec/parse_date/int_from_string_spec.rb (hunk 4 above)
Comment (ndushay) [reply spec/parse_date/int_from_string_spec.rb]: again, same - this was set up to work for our data at the time, and this is a straight port over.

File: spec/parse_date/int_from_string_spec.rb (hunk 5 above)
Comment (ndushay) [reply spec/parse_date/int_from_string_spec.rb]: you can't have everything.

File: spec/parse_date/int_from_string_spec.rb (hunk 7 above)
Comment (ndushay) [reply spec/parse_date/int_from_string_spec.rb]: direct port ... easier to just port over existing tests.  Plus, there's a lot of edge case functionality to test at individual private method level.

**Comment by ndushay** [comment]
Justin,

Thanks for your excellent eye for detail.  As I said in the description, "This particular PR ports over code from the stanford-mods gem for #year_int_from_date_str and for year_str_valid? methods from https://github.com/sul-dlss/stanford-mods/blob/master/lib/stanford-mods/date_parsing.rb "

As you often request with your straight code relocation PRs, I'm not sure this PR is the right place to address concerns with the code logic.  The existing code logic in stanford-mods was designed for accommodating SDR collections' date strings in a manner that would work for SearchWorks facets.   I didn't bring over some particularly squirrelly code that was too tied to SearchWorks (which had a sortable year as a string value, causing strange machinations for BC dates).

I shall add an example or two in the README for year_int_valid?.

I am certainly willing to change the code wrapper to be a class rather than a module.

File: README.md (hunk 6 above)
Comment (ndushay) [reply README.md]: done.

**Comment by ndushay** [comment]
@jcoyne assuming the tests pass (and they should) this is ready for re-review.

File: spec/parse_date/int_from_string_spec.rb (hunk 2 above)
Comment (jcoyne) [reply spec/parse_date/int_from_string_spec.rb]: The point I’m trying to get at is that these assumptions are strongly tied to Stanford metadata. These assumptions are not universal so we should make this “standard” and have Stanford mods gem have a way of making the Stanford specific assumptions. Baring that, some kind of configuration flags for 1900 + YY

File: lib/parse_date/int_from_string.rb [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+      matches&.to_s
+    end
+
+    # returns 4 digit year as String if we have a x/x/yy or x-x-yy pattern
+    #   note that these are the only 2 digit year patterns found in our actual date strings in MODS records
+    #   we use 20 as century digits unless it is greater than current year:
+    #   1/1/15  ->  2015
+    #   1/1/25  ->  1925
+    # @return [String, nil] 4 digit year (e.g. 1865, 0950) if orig_date_str matches pattern, nil otherwise
+    def sortable_year_for_yy(orig_date_str)
```
Comment (ndushay) [inline lib/parse_date/int_from_string.rb]: @jcoyne are these the comments you wanted?   I'm happy to have a follow on PR that adds this to the class method and to the README, if that would suffice.
